File: golang/3691_max_total_value/max_total_value.go
// Package main ...
package main

import (
	"container/heap"
	"fmt"
	"math"
)

type segTree struct {
	maxv, minv []int
	n          int
}

func newSegTree(nums []int) *segTree {
	n := len(nums)
	s := &segTree{
		maxv: make([]int, n*4),
		minv: make([]int, n*4),
		n:    n,
	}
	s.build(1, 0, n-1, nums)
	return s
}

func (s *segTree) build(node, l, r int, nums []int) {
	if l == r {
		s.maxv[node] = nums[l]
		s.minv[node] = nums[l]
		return
	}
	m := (l + r) / 2
	s.build(node*2, l, m, nums)
	s.build(node*2+1, m+1, r, nums)
	s.maxv[node] = max(s.maxv[node*2], s.maxv[node*2+1])
	s.minv[node] = min(s.minv[node*2], s.minv[node*2+1])
}

func (s *segTree) queryMax(node, l, r, ql, qr int) int {
	if ql <= l && r <= qr {
		return s.maxv[node]
	}
	m := (l + r) / 2
	res := math.MinInt
	if ql <= m {
		res = max(res, s.queryMax(node*2, l, m, ql, qr))
	}
	if qr > m {
		res = max(res, s.queryMax(node*2+1, m+1, r, ql, qr))
	}
	return res
}

func (s *segTree) queryMin(node, l, r, ql, qr int) int {
	if ql <= l && r <= qr {
		return s.minv[node]
	}
	m := (l + r) / 2
	res := math.MaxInt
	if ql <= m {
		res = min(res, s.queryMin(node*2, l, m, ql, qr))
	}
	if qr > m {
		res = min(res, s.queryMin(node*2+1, m+1, r, ql, qr))
	}
	return res
}

func maxTotalValue(nums []int, k int) int64 {
	n := len(nums)
	seg := newSegTree(nums)
	h := &hp{}
	for l := 0; l < n; l++ {
		heap.Push(h, tuple{seg.queryMax(1, 0, n-1, l, n-1) - seg.queryMin(1, 0, n-1, l, n-1), l, n - 1})
	}
	var ans int64 = 0
	for ; k > 0; k-- {
		t := heap.Pop(h).(tuple)
		ans += int64(t.val)
		if t.r > t.l {
			heap.Push(h, tuple{seg.queryMax(1, 0, n-1, t.l, t.r-1) - seg.queryMin(1, 0, n-1, t.l, t.r-1), t.l, t.r - 1})
		}
	}
	return ans
}

type (
	tuple struct{ val, l, r int }
	hp    []tuple
)

func (h hp) Len() int           { return len(h) }
func (h hp) Less(i, j int) bool { return h[i].val > h[j].val }
func (h hp) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *hp) Push(v any)        { *h = append(*h, v.(tuple)) }
func (h *hp) Pop() any          { a := *h; v := a[len(a)-1]; *h = a[:len(a)-1]; return v }

func main() {
	tests := []struct {
		nums []int
		k    int
		ans  int64
	}{
		{[]int{1, 3, 2}, 2, 4},
		{[]int{4, 2, 5, 1}, 3, 12},
	}

	for _, test := range tests {
		if maxTotalValue(test.nums, test.k) != test.ans {
			fmt.Println("test failed")
		}
	}
}

File: rust/3691_max_total_value/max_total_value.rs
struct Solution;

use std::collections::BinaryHeap;

struct SegTree {
    maxv: Vec<i32>,
    minv: Vec<i32>,
    n: usize,
}

impl SegTree {
    fn new(nums: &Vec<i32>) -> Self {
        let n = nums.len();
        let mut seg = SegTree {
            maxv: vec![0; n * 4],
            minv: vec![0; n * 4],
            n,
        };
        seg.build(1, 0, n - 1, nums);
        seg
    }

    fn build(&mut self, node: usize, l: usize, r: usize, nums: &Vec<i32>) {
        if l == r {
            self.maxv[node] = nums[l];
            self.minv[node] = nums[l];
            return;
        }
        let m = (l + r) / 2;
        self.build(node * 2, l, m, nums);
        self.build(node * 2 + 1, m + 1, r, nums);
        self.maxv[node] = self.maxv[node * 2].max(self.maxv[node * 2 + 1]);
        self.minv[node] = self.minv[node * 2].min(self.minv[node * 2 + 1]);
    }

    fn query_max(&self, node: usize, l: usize, r: usize, ql: usize, qr: usize) -> i32 {
        if ql <= l && r <= qr {
            return self.maxv[node];
        }
        let m = (l + r) / 2;
        let mut res = i32::MIN;
        if ql <= m {
            res = res.max(self.query_max(node * 2, l, m, ql, qr));
        }
        if qr > m {
            res = res.max(self.query_max(node * 2 + 1, m + 1, r, ql, qr));
        }
        res
    }

    fn query_min(&self, node: usize, l: usize, r: usize, ql: usize, qr: usize) -> i32 {
        if ql <= l && r <= qr {
            return self.minv[node];
        }
        let m = (l + r) / 2;
        let mut res = i32::MAX;
        if ql <= m {
            res = res.min(self.query_min(node * 2, l, m, ql, qr));
        }
        if qr > m {
            res = res.min(self.query_min(node * 2 + 1, m + 1, r, ql, qr));
        }
        res
    }
}

impl Solution {
    pub fn max_total_value(nums: Vec<i32>, k: i32) -> i64 {
        let n = nums.len();
        let seg = SegTree::new(&nums);
        let mut heap = BinaryHeap::new();
        for l in 0..n {
            let val = seg.query_max(1, 0, n - 1, l, n - 1) - seg.query_min(1, 0, n - 1, l, n - 1);
            heap.push((val, l, n - 1));
        }
        let mut ans: i64 = 0;
        let mut k = k as usize;
        while k > 0 {
            if let Some((val, l, r)) = heap.pop() {
                ans += val as i64;
                if r > l {
                    let new_val =
                        seg.query_max(1, 0, n - 1, l, r - 1) - seg.query_min(1, 0, n - 1, l, r - 1);
                    heap.push((new_val, l, r - 1));
                }
            }
            k -= 1;
        }
        ans
    }
}

fn main() {
    let tests = vec![(vec![1, 3, 2], 2, 4), (vec![4, 2, 5, 1], 3, 12)];

    for (nums, k, expected) in tests {
        assert_eq!(Solution::max_total_value(nums, k), expected);
    }
}
